File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/di/impl/dagger/modules/ReposModule.java
package com.duxtinto.intellij.plugin.github.codereviews.di.impl.dagger.modules;

import com.duxtinto.intellij.plugin.github.codereviews.di.scopes.ProjectScoped;
import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.RepositoriesDomainContract;
import com.duxtinto.intellij.plugin.github.codereviews.events.repos.GitChangeListener;
import com.duxtinto.intellij.plugin.github.codereviews.ide.acl.services.GitRepositoryFinderImpl;
import com.duxtinto.intellij.plugin.github.codereviews.ide.acl.services.GithubRepositoryFinderImpl;
import com.intellij.dvcs.repo.VcsRepositoryMappingListener;
import dagger.Binds;
import dagger.Module;

@Module
public abstract class ReposModule {
    @Binds
    @ProjectScoped
    public abstract RepositoriesDomainContract.GitRepositoryFinder provideGitRepositoryFinder(GitRepositoryFinderImpl gitRepositoryFinder);

    @Binds
    @ProjectScoped
    public abstract RepositoriesDomainContract.GithubRepositoryFinder provideGithubRepositoryFinder(GithubRepositoryFinderImpl githubRepositoryFinder);

    @Binds
    @ProjectScoped
    public abstract VcsRepositoryMappingListener provideVcsRepositoryMappingListener(GitChangeListener changeListener);
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/di/impl/dagger/modules/ReposModule.kt
package com.duxtinto.intellij.plugin.github.codereviews.di.impl.dagger.modules

import com.duxtinto.intellij.plugin.github.codereviews.di.scopes.ProjectScoped
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.ActionOnPullRequestInteractor
import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.RepositoriesDomainContract
import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.git.CheckoutBranchInteractor
import com.duxtinto.intellij.plugin.github.codereviews.ide.acl.services.git.GitRepositoryFinderImpl
import com.duxtinto.intellij.plugin.github.codereviews.ide.acl.services.github.repositories.GithubRepositoryFinderImpl
import dagger.Binds
import dagger.Module

@Module
abstract class ReposModule {
    @Binds
    @ProjectScoped
    abstract fun provideGitRepositoryFinder(gitRepositoryFinder: GitRepositoryFinderImpl): RepositoriesDomainContract.Git.RepositoryFinder

    @Binds
    @ProjectScoped
    abstract fun provideGithubRepositoryFinder(githubRepositoryFinder: GithubRepositoryFinderImpl): RepositoriesDomainContract.GithubRepositoryFinder

    @Binds
    @ProjectScoped
    abstract fun provideCheckoutBranchInteractor(interactor: CheckoutBranchInteractor): ActionOnPullRequestInteractor
}
